src/workerd/io/actor-sqlite.c++

```diff
@@ -351,9 +351,16 @@ kj::Promise<void> ActorSqlite::commitImpl(ActorSqlite::PrecommitAlarmState preco
 
   KJ_IF_SOME(pending, pendingCommit) {
     // If an earlier commitImpl() invocation is already in the process of updating precommit
-    // alarms but has not yet made the commitCallback() call, it should be OK to wait on it to
-    // perform the precommit alarm update and db commit for this invocation, too.
+    // alarms but has not yet made the commitCallback() call, wait on it to complete the db commit.
     co_await pending.addBranch();
+
+    // After the pending commit completes, check if our alarm state differs from what's scheduled.
+    // If so, we need to schedule it now, since the pending commit didn't know about our alarm state.
+    while (willFireEarlier(metadata.getAlarm(), alarmScheduledNoLaterThan)) {
+      KJ_LOG(WARNING, "NOSENTRY commitImpl merging new setAlarm with pending commit",
+          metadata.getAlarm(), alarmScheduledNoLaterThan);
+      co_await requestScheduledAlarm(metadata.getAlarm());
```

Comment:
Unrelated to this change, but related to the alarms problem you're trying to solve.

I read through the implementation of `requestScheduledAlarm` and saw this line:
```
  return hooks.scheduleRun(requestedTime).then([this, movingAlarmLater, requestedTime]() {
    if (!movingAlarmLater) {
      alarmScheduledNoLaterThan = requestedTime;
    }
  });
```
The internal implementation of `scheduleNow` can do all sorts of retries, so it's possible that if you have two alarm requests in quick succession, the first request finishes after the second request, which causes the first request's `then` closure to run after the second request's `then` closure.  What prevents this reordering from messing up `alarmScheduledNoLaterThan`?